python/lsst/ts/eas/glass_temperature_model.py

```diff
@@ -0,0 +1,162 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["GlassTemperatureModel"]
+
+import asyncio
+import logging
+import statistics
+from contextlib import AsyncExitStack
+
+from lsst.ts import salobj, utils
+from lsst.ts.xml.tables.m1m3 import find_thermocouple
+
+SAL_INDICES = (114, 115, 116, 117)
+N_THERMOCOUPLES = 146
+MAX_TIMESTAMP_AGE = 300
+
+
+class GlassTemperatureModel:
+    """A class to monitor the glass temperature for control of FCU fan speed.
+
+    This class provides monitoring of the glass temperature probes.
+    Fan speed scales linearly from 500 RPM at zero temperature difference
+    to 2000 RPM at ±1.0 °C difference. Values in between are interpolated,
+    and differences greater than or equal to 1.0 °C run at maximum speed.
+
+    The glass temperature probes publish to four different ESS CSCs, and
+    the measurements are multiplexed using the `sensorName` item in the
+    temperature telemetry.
+
+    Parameters
+    ----------
+    domain : `~lsst.ts.salobj.Domain`
+        A SAL domain object for obtaining remotes.
+    log : `~logging.Logger`
+        A logger for log messages.
+    """
+
+    def __init__(self, *, domain: salobj.Domain, log: logging.Logger) -> None:
+        self.domain = domain
+        self.log = log
+
+        self.monitor_start_event = asyncio.Event()
+        self.monitor_stop = asyncio.Event()
+
+        # A list of most recent temperature probe samples.
+        self.thermocouple_cache: list[tuple[float | None, float | None]] = [
+            (None, None)
+        ] * N_THERMOCOUPLES
+
+    async def temperature_callback(self, temperature: salobj.BaseMsgType) -> None:
+        """Callback for ESS.tel_temperature.
+
+        This function sorts out the data in the temperatureItem
+        array and stores it in the appropriate place in
+        `self.thermocouple_cache`. It is very much tailored
+        to the existing structure of the M1M3 thermal scanner
+        telemetry.
+
+        Parameters
+        ----------
+        temperature : `~lsst.ts.salobj.BaseMsgType`
+           A newly received temperature telemetry item.
+        """
+        sal_index = temperature.salIndex
+        sensor_name = temperature.sensorName  # Has format 'm1m3-ts-<index> <n>/6'
+        sequence_num = int(sensor_name.split()[1].split("/")[0])
+
+        n_channels = temperature.numChannels
+        temperatures = temperature.temperatureItem
+        temperatures = temperatures[:n_channels]
+
+        timestamp = temperature.timestamp
+
+        for i, temperature in enumerate(temperature.temperatureItem):
```

Comment:
This isn't as simple as that. One has to use https://github.com/lsst-ts/ts_xml/blob/develop/python/lsst/ts/xml/tables/m1m3/thermocouple_table.py find_thermocouple method to get "live", connected thermocouples. Better we transition the caching function to ts_m1m3_utils, and base the code on this. The current implementation will not filter garbage we send, and will produce results we cannot trust.